torch/lowrank.py

```diff
@@ -0,0 +1,249 @@
+
+import torch
+from . import _linalg_utils as _utils 
+
+
+def get_approximate_basis(A, q, niter=2, M=None):
+    """Return tensor :math:`Q` with q orthonormal columns such that
+    :math:`Q Q^H A` approximates :math:`A`. If :math:`M` is specified,
+    then `Q` is such that :math:`Q Q^H (A - M)` approximates :math:`A
+    - M`.
+
+    .. note:: The implementation is based on the Algorithm 4.4 from
+              Halko et al, 2009.
+
+    .. note:: For an adequate approximation of a k-rank matrix A,
+              where k is not known in advance but could be estimated,
+              the number of Q columns q can be choosen according to
+              the following criteria: in general, :math:`k <= q <=
+              min(2*k, m, n)`. For large low-rank matrices, take
+              :math:`q = k + 5..10`.  If k is relatively small
+              compared to :math:`min(m, n)`, choosing :math:`q = k +
+              0..2` may be sufficient.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int): the dimension of subspace spanned by Q columns.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer. In most cases, the default
+                               value 2 is more than enough.
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    dtype = _utils.get_floating_dtype(A)
+    matmul = _utils.get_matmul(A)
+
+    R = _utils.uniform(low=-1.0, high=1.0, size=(n, q),
+                       dtype=dtype, device=A.device)
+
+    A_H = _utils.transjugate(A)
+    if M is None:
+        (Q, _) = matmul(A, R).qr()
+        for i in range(niter):
+            (Q, _) = matmul(A_H, Q).qr()
+            (Q, _) = matmul(A, Q).qr()
+    else:
+        matmul2 = _utils.get_matmul(M)
+        M_H = _utils.transjugate(M)
+        (Q, _) = (matmul(A, R) - matmul2(M, R)).qr()
+        for i in range(niter):
+            (Q, _) = (matmul(A_H, Q) - matmul2(M_H, Q)).qr()
+            (Q, _) = (matmul(A, Q) - matmul2(M, Q)).qr()
+
+    return Q
+
+
+def svd(A, q=6, niter=2, M=None):
+    """Return the singular value decomposition ``(U, S, V)`` of a matrix,
+    batches of matrices, or a sparse matrix :math:`A` such that
+    :math:`A \approx U diag(S) V^T`. In case :math:`M` is given, then
+    SVD is computed for the matrix :math:`A - M`.
```

Comment:
The interface suggests that it is a full-rank SVD, but then there is this approximation. The approximation is in the spectral norm, right?

Reply:
I presume the method is not the best candidate when `A` is actually (nearly)full-rank, right? If not, then maybe it is also worth mentioning this in the description.

Reply:
Sorry, just read the note on low-rank input. My apologies

Reply:
There is a faster, but less accurate algorithm, described in Algorithm 5.2. It is a slight modification of Algorithm 5.1. Maybe would be desirable to have it implemented. In most data analysis applications speed is more important than accuracy.

Reply:
There is also an option to return a k-rank approximation given that overparameterized matrix Q with `rank(Q) = q > k` (subsection 5.2). I think it makes sense to implement it, and it is required for Algorithm 5.2. The reason being is that a user may naively just drop redundant vectors and obtain a factorization with higher error. This postprocessing does not require computing any matrix-matrix products, so is also extremely fast.

Reply:
For speed, one should use `torch.svd`.

Halko et al, 2009, propose six algorithms for low-rank SVD problems. Here only the direct method is implemented as MVP.

Reply:
Ok, will add to the proposal.